WildflyIT.testTransportClient failed in CI

Failed in CI today - https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=amazon/911/console
```
  2> NOTE: Linux 4.4.44-39.55.amzn1.x86_64 amd64/Oracle Corporation 1.8.0_121 (64-bit)/cpus=4,threads=1,free=483862184,total=514850816
  2> NOTE: All tests run in this JVM: [WildflyIT]
   > Expected: <201>
   >      but: was <400>
   > 	at __randomizedtesting.SeedInfo.seed([990E732BDE24D8EF:F5B837CA4E02164A]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.wildfly.WildflyIT.testTransportClient(WildflyIT.java:78)
   > 	at java.lang.Thread.run(Thread.java:745)
Completed [1/1] in 1.11s, 1 test, 1 failure <<< FAILURES!

[ant:junit4] JVM J0: stderr was not empty, see: /var/lib/jenkins/workspace/elastic+elasticsearch+master+multijob-unix-compatibility/os/amazon/qa/wildfly/build/testrun/integTestRunner/temp/junit4-J0-20170502_094849_4827938408300352232738.syserr
Tests with failures:
  - org.elasticsearch.wildfly.WildflyIT.testTransportClient
```

It seems to be the first failure with this particular error. However this test failed 6 days ago with a different, but possibly related error:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-intake/1073/console

```
2> NOTE: reproduce with: ant test  -Dtestcase=WildflyIT -Dtests.method=testTransportClient -Dtests.seed=88966EDD55EC9F58 -Dtests.locale=hr -Dtests.timezone=America/St_Kitts -Dtests.asserts=true -Dtests.file.encoding=UTF8
ERROR   0.39s | WildflyIT.testTransportClient <<< FAILURES!
   > Throwable #1: org.apache.http.client.ClientProtocolException
  2> NOTE: test params are: codec=Asserting(Lucene62): {}, docValues:{}, maxPointsInLeafNode=677, maxMBSortInHeap=6.504225533270213, sim=RandomSimilarity(queryNorm=false,coord=yes): {}, locale=hr, timezone=America/St_Kitts
  2> NOTE: Linux 4.4.44-39.55.amzn1.x86_64 amd64/Oracle Corporation 1.8.0_121 (64-bit)/cpus=4,threads=1,free=487412120,total=514850816
  2> NOTE: All tests run in this JVM: [WildflyIT]
   > 	at __randomizedtesting.SeedInfo.seed([88966EDD55EC9F58:E4202A3CC5CA51FD]:0)
   > 	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
   > 	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
   > 	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
   > 	at org.elasticsearch.wildfly.WildflyIT.testTransportClient(WildflyIT.java:77)
   > 	at java.lang.Thread.run(Thread.java:745)
   > Caused by: org.apache.http.ProtocolException: The server failed to respond with a valid HTTP response
   > 	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:151)
   > 	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57)
   > 	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
   > 	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
   > 	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:167)
   > 	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
   > 	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
   > 	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:271)
   > 	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
   > 	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
   > 	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
   > 	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
   > 	... 39 more
Completed [1/1] in 1.09s, 1 test, 1 error <<< FAILURES!

```

From the logs, HTTP services did not fully startup:

```
2017-05-02 09:48:43,993 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-2) MSC000001: Failed to start service jboss.undertow.listener.default: org.jboss.msc.service.StartException in service jboss.undertow.listener.default: Could not start http listener
	at org.wildfly.extension.undertow.ListenerService.start(ListenerService.java:142)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.xnio.nio.NioXnioWorker.createTcpConnectionServer(NioXnioWorker.java:190)
	at org.xnio.XnioWorker.createStreamConnectionServer(XnioWorker.java:243)
	at org.wildfly.extension.undertow.HttpListenerService.startListening(HttpListenerService.java:126)
	at org.wildfly.extension.undertow.ListenerService.start(ListenerService.java:138)
	... 5 more
```

This is a known issue that can happen with the Wildfly tests because right now we can not use ephemeral ports. I have opened a PR against Wildfly (wildfly/wildfly#9943) so that we can move to ephemeral ports but there's a follow-up PR needed against another Wildfly component before we can fully implement this (we need changes on our side too). This is very low on my priority list.

--- a/qa/wildfly/build.gradle
+++ b/qa/wildfly/build.gradle
@@ -29,14 +29,10 @@ apply plugin: 'war'
 apply plugin: 'elasticsearch.build'
 apply plugin: 'elasticsearch.rest-test'
 
-final String wildflyVersion = '10.0.0.Final'
+final String wildflyVersion = '11.0.0.Final'
 final String wildflyDir = "${buildDir}/wildfly"
 final String wildflyInstall = "${buildDir}/wildfly/wildfly-${wildflyVersion}"
-// TODO: use ephemeral ports
-final int portOffset = 30000
-final int managementPort = 9990 + portOffset
-// we skip these tests on Windows so we do not need to worry about compatibility here
-final String stopWildflyCommand = "${wildflyInstall}/bin/jboss-cli.sh --controller=localhost:${managementPort} --connect command=shutdown"
+int managementPort
 
 repositories {
     mavenCentral()
@@ -114,31 +110,53 @@ task enableIPv6 {
 
 task startWildfly {
     dependsOn enableIPv6, writeElasticsearchProperties
-    doFirst {
-        // we skip these tests on Windows so we do no need to worry about compatibility here
-        final File script = new File(project.buildDir, "wildfly/wildfly.killer.sh")
-        script.setText(
-                ["function shutdown {",
-                 "  ${stopWildflyCommand}",
-                 "}",
-                 "trap shutdown EXIT",
-                 // will wait indefinitely for input, but we never pass input, and the pipe is only closed when the build dies
-                 "read line"].join('\n'), 'UTF-8')
-        final ProcessBuilder pb = new ProcessBuilder("bash", script.absolutePath)
-        pb.start()
-    }
     doLast {
         // we skip these tests on Windows so we do no need to worry about compatibility here
-        final ProcessBuilder pb =
-                new ProcessBuilder("${wildflyInstall}/bin/standalone.sh", "-Djboss.socket.binding.port-offset=${portOffset}")
-        final Process process = pb.start()
+        final ProcessBuilder wildfly = new ProcessBuilder(
+                "${wildflyInstall}/bin/standalone.sh",
+                "-Djboss.http.port=0",
+                "-Djboss.https.port=0",
+                "-Djboss.management.http.port=0")
+        final Process process = wildfly.start()
         new BufferedReader(new InputStreamReader(process.getInputStream())).withReader { br ->
             String line
+            int httpPort = 0
             while ((line = br.readLine()) != null) {
-                if (line.matches(".*WildFly Full \\d+\\.\\d+\\.\\d+\\.Final \\(WildFly Core \\d+\\.\\d+\\.\\d+\\.Final\\) started.*")) {
+                if (line.matches('.*Undertow HTTP listener default listening on .*:\\d+$')) {
+                    final int index = line.lastIndexOf(":")
+                    assert index >= 0
+                    httpPort = Integer.parseInt(line.substring(index + 1))
+                    // set this system property so the test runner knows the port Wildfly is listening for HTTP requests on
+                    integTestRunner.systemProperty("tests.jboss.http.port", httpPort)
+                } else if (line.matches('.*Http management interface listening on http://.*:\\d+/management$')) {
+                    final int colonIndex = line.lastIndexOf(":")
+                    assert colonIndex >= 0
+                    final int slashIndex = line.lastIndexOf("/")
+                    assert slashIndex >= 0
+                    managementPort = Integer.parseInt(line.substring(colonIndex + 1, slashIndex))
+
+                    /*
+                     * As soon as we know the management port, we fork a process that will ensure the Wildfly process is killed if we
+                     * teardown abnormally. We skip these tests on Windows so we do no need to worry about CLI compatibility here.
+                     */
+                    final File script = new File(project.buildDir, "wildfly/wildfly.killer.sh")
+                    script.setText(
+                            ["function shutdown {",
+                             "  ${wildflyInstall}/bin/jboss-cli.sh --controller=localhost:${-> managementPort} --connect command=shutdown",
+                             "}",
+                             "trap shutdown EXIT",
+                             // will wait indefinitely for input, but we never pass input, and the pipe is only closed when the build dies
+                             "read line\n"].join('\n'), 'UTF-8')
+                    final ProcessBuilder killer = new ProcessBuilder("bash", script.absolutePath)
+                    killer.start()
+
+                } else if (line.matches(".*WildFly Full \\d+\\.\\d+\\.\\d+\\.Final \\(WildFly Core \\d+\\.\\d+\\.\\d+\\.Final\\) started.*")) {
                     break
                 }
             }
+
+            assert httpPort > 0
+            assert managementPort > 0
         }
     }
 }
@@ -147,13 +165,14 @@ task configureTransportClient(type: LoggedExec) {
     dependsOn startWildfly
     // we skip these tests on Windows so we do not need to worry about compatibility here
     commandLine "${wildflyInstall}/bin/jboss-cli.sh",
-            "--controller=localhost:${managementPort}",
+            "--controller=localhost:${-> managementPort}",
             "--connect",
             "--command=/system-property=elasticsearch.properties:add(value=\${jboss.server.config.dir}/elasticsearch.properties)"
 }
 
 task stopWildfly(type: LoggedExec) {
-    commandLine stopWildflyCommand.split(' ')
+    // we skip these tests on Windows so we do no need to worry about CLI compatibility here
+    commandLine "${wildflyInstall}/bin/jboss-cli.sh", "--controller=localhost:${-> managementPort}", "--connect", "command=shutdown"
 }
 
 if (!Os.isFamily(Os.FAMILY_WINDOWS) && project.rootProject.ext.javaVersion == JavaVersion.VERSION_1_8) {

--- a/qa/wildfly/build.gradle
+++ b/qa/wildfly/build.gradle
@@ -123,12 +123,14 @@ task startWildfly {
             int httpPort = 0
             while ((line = br.readLine()) != null) {
                 if (line.matches('.*Undertow HTTP listener default listening on .*:\\d+$')) {
+                    assert httpPort == 0
                     final int index = line.lastIndexOf(":")
                     assert index >= 0
                     httpPort = Integer.parseInt(line.substring(index + 1))
                     // set this system property so the test runner knows the port Wildfly is listening for HTTP requests on
                     integTestRunner.systemProperty("tests.jboss.http.port", httpPort)
                 } else if (line.matches('.*Http management interface listening on http://.*:\\d+/management$')) {
+                    assert managementPort
                     final int colonIndex = line.lastIndexOf(":")
                     assert colonIndex >= 0
                     final int slashIndex = line.lastIndexOf("/")

--- a/qa/wildfly/src/test/java/org/elasticsearch/wildfly/WildflyIT.java
+++ b/qa/wildfly/src/test/java/org/elasticsearch/wildfly/WildflyIT.java
@@ -53,7 +53,8 @@ public class WildflyIT extends LuceneTestCase {
         try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
             final String str = String.format(
                     Locale.ROOT,
-                    "http://localhost:38080/wildfly-%s%s/transport/employees/1",
+                    "http://localhost:%d/wildfly-%s%s/transport/employees/1",
+                    Integer.parseInt(System.getProperty("tests.jboss.http.port")),
                     Version.CURRENT,
                     Build.CURRENT.isSnapshot() ? "-SNAPSHOT" : "");
             final HttpPut put = new HttpPut(new URI(str));

